# Análisis de Finanzas y Riesgo Crediticio 26.01.2026

## Librerias

In [7]:
import pandas as pd
import numpy as np

import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats import kruskal
from sqlalchemy import create_engine



## 1. Creamos conexión

In [8]:

# Crear el engine de conexión
engine = create_engine(
    "mysql+mysqlconnector://Equipo21:E1q2u3i4p5o21@212.227.90.6/Equip_21"
)

# Cargar la view en un DataFrame
df = pd.read_sql("SELECT * FROM BANK_marketing_deduplicated;", engine)
df.head()

,id,age,job,marital,education,credit_default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,deposit
0,1,59,admin.,married,secondary,0,2343,1,0,low_quality,5,may,1042,1,-1,0,unknown,1
1,2,59,admin.,married,secondary,0,2343,1,0,low_quality,5,may,1042,1,-1,0,unknown,1
2,3,56,admin.,married,secondary,0,45,0,0,low_quality,5,may,1467,1,-1,0,unknown,1
3,4,41,technician,married,secondary,0,1270,1,0,low_quality,5,may,1389,1,-1,0,unknown,1
4,5,55,services,married,secondary,0,2476,1,0,low_quality,5,may,579,1,-1,0,unknown,1


In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10988 entries, 0 to 10987
Data columns (total 18 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   id              10988 non-null  int64 
 1   age             10988 non-null  int64 
 2   job             10988 non-null  object
 3   marital         10988 non-null  object
 4   education       10988 non-null  object
 5   credit_default  10988 non-null  int64 
 6   balance         10988 non-null  int64 
 7   housing         10988 non-null  int64 
 8   loan            10988 non-null  int64 
 9   contact         10988 non-null  object
 10  day             10988 non-null  int64 
 11  month           10988 non-null  object
 12  duration        10988 non-null  int64 
 13  campaign        10988 non-null  int64 
 14  pdays           10988 non-null  int64 
 15  previous        10988 non-null  int64 
 16  poutcome        10988 non-null  object
 17  deposit         10988 non-null  int64 
dtypes: int

## 2. Análisis de variables relevantes

### 2.1. Balance

In [5]:
# Estadísticas descriptivas ampliadas de balance
df["balance"].describe(percentiles=[0.01, 0.05, 0.10, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99])


count    10988.000000
mean      1538.235985
std       3239.869418
min      -6847.000000
1%        -522.000000
5%         -51.000000
10%          0.000000
25%        125.000000
50%        556.000000
75%       1723.000000
90%       3913.300000
95%       6032.850000
99%      13338.520000
max      81204.000000
Name: balance, dtype: float64

### 2.2. Credit default

In [19]:
# Conteo y %
df["credit_default"].value_counts(), df["credit_default"].value_counts(normalize=True) * 100

(credit_default
 0    10827
 1      161
 Name: count, dtype: int64,
 credit_default
 0    98.534765
 1     1.465235
 Name: proportion, dtype: float64)

### 2.3. Productos financieros - loan & housing

In [7]:
# Count y pct por producto
def counts_with_pct(series):
    counts = series.value_counts(dropna=False)
    pct = series.value_counts(dropna=False, normalize=True).mul(100).round(2)
    return pd.DataFrame({"count": counts, "pct": pct})

display(counts_with_pct(df["loan"]))
display(counts_with_pct(df["housing"]))

,count,pct
loan,,
0,9557,86.98
1,1431,13.02


,count,pct
housing,,
0,5804,52.82
1,5184,47.18


## 3. Análisis avg_balance 

### 3.1. Balance distribution

La distribución del balance es fuertemente asimétrica y presenta numerosos outliers, por lo que se utilizan métodos no paramétricos en el análisis.


In [8]:
df["balance"].describe(
    percentiles=[0.01, 0.05, 0.25, 0.50, 0.75, 0.95, 0.99]
).round(2)


count    10988.00
mean      1538.24
std       3239.87
min      -6847.00
1%        -522.00
5%         -51.00
25%        125.00
50%        556.00
75%       1723.00
95%       6032.85
99%      13338.52
max      81204.00
Name: balance, dtype: float64

### 3.2. avg_balance by loan & housing


In [9]:
df["product_group"] = (
    "loan=" + df["loan"].astype(str) +
    ", housing=" + df["housing"].astype(str)
)


In [15]:
# Tabla descriptiva por cada combinación de loan & housing
summary_table = (
    df
    .groupby("product_group")["balance"]
    .agg(
        count="count",
        mean="mean",
        median="median",
        q25=lambda x: x.quantile(0.25),
        q75=lambda x: x.quantile(0.75)
    )
    .reset_index()
    .round(2)
)

summary_table


,product_group,count,mean,median,q25,q75
0,"loan=0, housing=0",5191,1877.98,703.0,191.00,2202.50
1,"loan=0, housing=1",4366,1365.55,501.5,105.00,1504.00
2,"loan=1, housing=0",613,912.61,290.0,28.00,905.00
3,"loan=1, housing=1",818,772.79,300.5,1.25,991.75


### 3.3. Test Estadísticos - Kruskal–Wallis

Los clientes con préstamos y/o hipotecas presentan un saldo medio significativamente más bajo que los clientes sin estos productos.
La diferencia es estadísticamente significativa (Kruskal–Wallis, p < 0.001) y es más marcada en clientes con préstamos


El test de Kruskal–Wallis se utiliza por la naturaleza asimétrica del balance y la presencia de outliers.

**H₀:** No hay diferencias significativas en el saldo entre los distintos tipos de clientes.  
**H₁:** Existen diferencias significativas en el saldo entre al menos dos tipos de clientes.


In [10]:

groups = [
    df.loc[df["product_group"] == "loan=0, housing=0", "balance"],
    df.loc[df["product_group"] == "loan=0, housing=1", "balance"],
    df.loc[df["product_group"] == "loan=1, housing=0", "balance"],
    df.loc[df["product_group"] == "loan=1, housing=1", "balance"]
]

stat, p_value = kruskal(*groups)
stat, p_value


(np.float64(317.5058614927561), np.float64(1.6167972181305572e-68))

In [11]:
# Tabla resumen del test Kruskal–Wallis
kruskal_results = pd.DataFrame({
    "Test": ["Kruskal–Wallis"],
    "H statistic": [round(stat, 2)],
    "p-value": [f"{p_value:.2e}"],
    "Significativo (alpha=0.05)": ["Si" if p_value < 0.05 else "No"]
})

kruskal_results


,Test,H statistic,p-value,Significativo (alpha=0.05)
0,Kruskal–Wallis,317.51,1.62e-68,Si


#### 3.3.1. Análisis por grado de vinculación (número de productos contratados)



In [12]:
df["num_products"] = df["loan"] + df["housing"]

df.groupby("num_products")["balance"].median()



num_products
0    703.0
1    471.0
2    300.5
Name: balance, dtype: float64

In [13]:

groups_np = [
    df.loc[df["num_products"] == 0, "balance"],
    df.loc[df["num_products"] == 1, "balance"],
    df.loc[df["num_products"] == 2, "balance"]
]

stat_np, p_value_np = kruskal(*groups_np)
stat_np, p_value_np


(np.float64(274.1186103127201), np.float64(2.9915762526158446e-60))

In [14]:
pd.DataFrame({
    "Test": ["Kruskal–Wallis (num_products)"],
    "H statistic": [round(stat_np, 2)],
    "p-value": [f"{p_value_np:.2e}"],
    "Significativo (alpha=0.05)": ["Sí" if p_value_np < 0.05 else "No"]
})


,Test,H statistic,p-value,Significativo (alpha=0.05)
0,Kruskal–Wallis (num_products),274.12,2.99e-60,Sí


## 4. Análisis credit_default 

### 4.1. Dataframe de pct_default por combinaciones de loan & housing

In [10]:
# Df resumen con tasas de impago
df_default = (
    df.groupby(["loan", "housing"])
      .agg(
          pct_default = ("credit_default", "mean"),
          n_clients = ("id", "count")
      )
      .reset_index()
)

# Convertir prob_impago a porcentaje
df_default["pct_default"] = df_default["pct_default"] * 100

df_default

,loan,housing,pct_default,n_clients
0,0,0,0.866885,5191
1,0,1,1.397160,4366
2,1,0,5.383361,613
3,1,1,2.689487,818


### 4.2. Tests Estadísticos - Z-Test de proporciones


In [11]:
from statsmodels.stats.proportion import proportions_ztest

# --- Definimos los grupos A, B, C, D según el DataFrame ---

A = df[(df["loan"]==0) & (df["housing"]==0)]   # Sin préstamo, sin hipoteca
B = df[(df["loan"]==0) & (df["housing"]==1)]   # Sin préstamo, con hipoteca
C = df[(df["loan"]==1) & (df["housing"]==0)]   # Con préstamo, sin hipoteca
D = df[(df["loan"]==1) & (df["housing"]==1)]   # Con préstamo, con hipoteca


# --- Función auxiliar para hacer el z-test entre dos grupos ---
def ztest_groups(g1, g2):
    default = [g1["credit_default"].sum(), g2["credit_default"].sum()]
    n = [len(g1), len(g2)]
    stat, pval = proportions_ztest(default, n)
    return stat, pval

### 4.2.1  Pregunta 1: ¿Tener préstamo aumenta el riesgo?

**C vs A (solo préstamo vs ningún producto)**
- H₀: Las tasas de impago son iguales entre los clientes que tienen solo préstamo y los que no tienen ningún producto.
- H₁: Las tasas de impago son distintas entre los clientes que tienen solo préstamo y los que no tienen ningún producto.

**D vs B (préstamo+hipoteca vs solo hipoteca)**
- H₀: Las tasas de impago son iguales entre los clientes que tienen tanto préstamo como hipoteca y los que tienen solo hipoteca.
- H₁: Las tasas de impago son distintas entre los clientes que tienen tanto préstamo como hipoteca y los que tienen solo hipoteca.

In [14]:
alpha = 0.05

# C vs A → loan=1,housing=0 vs loan=0,housing=0
stat_CA, p_CA = ztest_groups(C, A)

# D vs B → loan=1,housing=1 vs loan=0,housing=1
stat_DB, p_DB = ztest_groups(D, B)

# Función para decidir si se rechaza H0
def decision(p):
    return "Rechazamos H₀ (diferencia significativa)" if p < alpha else "No rechazamos H₀"

# Crear tabla resumen
results = pd.DataFrame({
    "Comparación": [
        "C vs A (solo préstamo vs ningún producto)",
        "D vs B (préstamo+hipoteca vs solo hipoteca)"
    ],
    "Estadístico Z": [round(stat_CA, 3), round(stat_DB, 3)],
    "p-value": [round(p_CA, 6), round(p_DB, 6)],
    "Decisión (alpha = 0.05)": [decision(p_CA), decision(p_DB)]
})

print("\nResultados Z-test entre grupos:\n")
print(results.to_string(index=False))


Resultados Z-test entre grupos:

                                Comparación  Estadístico Z  p-value                  Decisión (alpha = 0.05)
  C vs A (solo préstamo vs ningún producto)          9.184 0.000000 Rechazamos H₀ (diferencia significativa)
D vs B (préstamo+hipoteca vs solo hipoteca)          2.702 0.006883 Rechazamos H₀ (diferencia significativa)


**Conclusión: Rechazamos todas las hipótesis nulas**

Los clientes con préstamo personal (loan=1) tienen una probabilidad de impago significativamente mayor que los clientes sin ningún producto.
El préstamo personal es un factor de riesgo muy fuerte.

Añadir un préstamo personal a un cliente que ya tiene hipoteca incrementa significativamente el riesgo de impago.
La hipoteca sola no es tan problemática, pero cuando se combina con un préstamo, el riesgo sube.

### 4.2.2. Pregunta 2: ¿Tener hipoteca aumenta el riesgo?

**B vs A (solo hipoteca vs ningún producto)**
- H₀: Las tasas de impago son iguales entre los clientes que tienen solo hipoteca y los que no tienen ningún producto.
- H₁: Las tasas de impago son distintas entre los clientes que tienen solo hiipoteca y los que no tienen ningún producto.

**D vs C (préstamo+hipoteca vs solo préstamo)**
- H₀: Las tasas de impago son iguales entre los clientes que tienen tanto préstamo como hipoteca y los que tienen solo préstamo.
- H₁: Las tasas de impago son distintas entre los clientes que tienen tanto préstamo como hipoteca y los que tienen solo préstamo.

In [16]:
alpha = 0.05

# B vs A → solo hipoteca vs ningún producto
stat_BA, p_BA = ztest_groups(B, A)

# D vs C → préstamo+hipoteca vs solo préstamo
stat_DC, p_DC = ztest_groups(D, C)

# Función para decidir si se rechaza H0
def decision(p):
    return "Rechazamos H₀ (diferencia significativa)" if p < alpha else "No rechazamos H₀"

# Crear tabla resumen
results_extra = pd.DataFrame({
    "Comparación": [
        "B vs A (solo hipoteca vs ningún producto)",
        "D vs C (préstamo+hipoteca vs solo préstamo)"
    ],
    "Estadístico Z": [round(stat_BA, 3), round(stat_DC, 3)],
    "p-value": [round(p_BA, 6), round(p_DC, 6)],
    "Decisión (alpha = 0.05)": [decision(p_BA), decision(p_DC)]
})

print("\nResultados Z-test adicionales:\n")
print(results_extra.to_string(index=False))


Resultados Z-test adicionales:

                                Comparación  Estadístico Z  p-value                  Decisión (alpha = 0.05)
  B vs A (solo hipoteca vs ningún producto)          2.466 0.013675 Rechazamos H₀ (diferencia significativa)
D vs C (préstamo+hipoteca vs solo préstamo)         -2.623 0.008714 Rechazamos H₀ (diferencia significativa)


**Conclusión: Rehazamos todas las hipótesis nulas**

Los clientes con solo hipoteca tienen un riesgo de impago significativamente mayor que los clientes sin productos.
 - La hipoteca sí aumenta el riesgo, pero mucho menos que el préstamo personal.
 - Es un efecto real, pero pequeño comparado con el impacto del préstamo.

Entre los clientes que ya tienen un préstamo personal, añadir una hipoteca reduce significativamente el riesgo de impago.
 - Esto confirma queÇ el grupo más riesgoso es C (solo préstamo).
 - El grupo D (préstamo + hipoteca) tiene riesgo alto, pero menor que C.

### 4.2.3. Pregunta 3: ¿El grupo préstamo + hipoteca es el más riesgoso?

**D vs C (préstamo+hipoteca vs solo préstamo)**
- H₀: Las tasas de impago son iguales entre los clientes que tienen tanto préstamo como hipoteca y los que tienen solo préstamo.
- H₁: Las tasas de impago son distintas entre los clientes que tienen tanto préstamo como hipoteca y los que tienen solo préstamo.

**D vs A (préstamo+hipoteca vs ningún producto)**
- H₀: Las tasas de impago son iguales entre los clientes que tienen tanto préstamo como hipoteca y los que no tienen ningún producto.
- H₁: Las tasas de impago son distintas entre los clientes que tienen tanto préstamo como hipoteca y los no que tienen ningún producto.

**D vs B (préstamo+hipoteca vs solo hipoteca)**
- H₀: Las tasas de impago son iguales entre los clientes que tienen tanto préstamo como hipoteca y los que tienen solo hipoteca.
- H₁: Las tasas de impago son distintas entre los clientes que tienen tanto préstamo como hipoteca y los que tienen solo hipoteca.

In [15]:
alpha = 0.05

# D vs C → préstamo+hipoteca vs solo préstamo
stat_DC, p_DC = ztest_groups(D, C)

# D vs A → préstamo+hipoteca vs ningún producto
stat_DA, p_DA = ztest_groups(D, A)

# D vs B → préstamo+hipoteca vs solo hipoteca
stat_DB, p_DB = ztest_groups(D, B)

# Función para decidir si se rechaza H0
def decision(p):
    return "Rechazamos H₀ (diferencia significativa)" if p < alpha else "No rechazamos H₀"

# Crear tabla resumen
results2 = pd.DataFrame({
    "Comparación": [
        "D vs C (préstamo+hipoteca vs solo préstamo)",
        "D vs A (préstamo+hipoteca vs ningún producto)",
        "D vs B (préstamo+hipoteca vs solo hipoteca)"
    ],
    "Estadístico Z": [round(stat_DC, 3), round(stat_DA, 3), round(stat_DB, 3)],
    "p-value": [round(p_DC, 6), round(p_DA, 6), round(p_DB, 6)],
    "Decisión (alpha = 0.05)": [decision(p_DC), decision(p_DA), decision(p_DB)]
})

print("\nResultados Z-test entre grupos:\n")
print(results2.to_string(index=False))


Resultados Z-test entre grupos:

                                  Comparación  Estadístico Z  p-value                  Decisión (alpha = 0.05)
  D vs C (préstamo+hipoteca vs solo préstamo)         -2.623 0.008714 Rechazamos H₀ (diferencia significativa)
D vs A (préstamo+hipoteca vs ningún producto)          4.614 0.000004 Rechazamos H₀ (diferencia significativa)
  D vs B (préstamo+hipoteca vs solo hipoteca)          2.702 0.006883 Rechazamos H₀ (diferencia significativa)


**Conclusión: Rechazamos todas las hipótesis nulas**
  
Entre los clientes con préstamo personal, tener hipoteca reduce el riesgo. 
 - El grupo más riesgoso de todos es el de clientes con solo préstamo (C).

Los clientes con préstamo + hipoteca (D) tienen un riesgo de impago mucho mayor que los clientes sin productos (A). 
 - Es claramente un grupo de riesgo elevado comparado con el baseline.

Añadir un préstamo personal a un cliente con hipoteca incrementa significativamente el riesgo. 
 - La hipoteca sola (B) tiene riesgo moderado.
 - La combinación préstamo + hipoteca (D) empeora ese riesgo.

### 4.3. Tests Estadísticos - Chi-cuadrado de independencia

#### 4.3.1. Existe relación entre las variables loan vs credit default?

- H₀: loan y credit_default son independientes
- H₁: loan y credit_default están asociados

In [17]:
from scipy.stats import chi2_contingency

alpha = 0.05

# Tabla de contingencia: loan vs impago
loan_table = pd.crosstab(df["loan"], df["credit_default"])
chi2_loan, p_loan, dof_loan, expected_loan = chi2_contingency(loan_table)

# Función para decidir si se rechaza H0
def decision(p):
    return "Rechazamos H₀ (variables asociadas)" if p < alpha else "No rechazamos H₀ (variables independientes)"

# Crear DataFrame para mostrar resultados
loan_results = pd.DataFrame({
    "Métrica": ["Chi-cuadrado", "p-value", "Grados de libertad", "Decisión (alpha = 0.05)"],
    "Valor": [round(chi2_loan, 3), round(p_loan, 6), dof_loan, decision(p_loan)]
})

print("\n Chi-cuadrado: LOAN vs IMPAGO\n")
print(loan_results.to_string(index=False))

print("\n Tabla observada:")
print(loan_table.to_string())

print("\n Frecuencias esperadas:")
print(pd.DataFrame(expected_loan,
                   index=loan_table.index,
                   columns=loan_table.columns).round(2).to_string())


 Chi-cuadrado: LOAN vs IMPAGO

                Métrica                               Valor
           Chi-cuadrado                              62.574
                p-value                                 0.0
     Grados de libertad                                   1
Decisión (alpha = 0.05) Rechazamos H₀ (variables asociadas)

 Tabla observada:
credit_default     0    1
loan                     
0               9451  106
1               1376   55

 Frecuencias esperadas:
credit_default        0       1
loan                           
0               9416.97  140.03
1               1410.03   20.97


**Conlusión: Rechazamos la hipótesis nula**
- Existe una relación estadísticamente muy fuerte entre tener un préstamo personal y el impago.|
- El p‑value ≈ 0 indica que esta relación no es fruto del azar.
- Los clientes con préstamo presentan muchos más impagos de los esperados si no hubiera relación entre ambas variables.
- Los clientes sin préstamo muestran muchos menos impagos de los esperados, lo que refuerza la asociación.

El préstamo personal es un factor de riesgo crítico y explica una parte sustancial del comportamiento de impago en la cartera.


#### 4.3.2. Existe relación entre las variables housing vs credit default?

- H₀: housing y credit_default son independientes
- H₁: housing y credit_default están asociados

In [18]:
from scipy.stats import chi2_contingency

alpha = 0.05

# Tabla de contingencia: housing vs impago
housing_table = pd.crosstab(df["housing"], df["credit_default"])
chi2_housing, p_housing, dof_housing, expected_housing = chi2_contingency(housing_table)

# Función para decidir si se rechaza H0
def decision(p):
    return "Rechazamos H₀ (variables asociadas)" if p < alpha else "No rechazamos H₀ (variables independientes)"

# Crear DataFrame para mostrar resultados
housing_results = pd.DataFrame({
    "Métrica": ["Chi-cuadrado", "p-value", "Grados de libertad", "Decisión (alpha = 0.05)"],
    "Valor": [round(chi2_housing, 3), round(p_housing, 6), dof_housing, decision(p_housing)]
})

print("\n Chi-cuadrado: HOUSING vs IMPAGO\n")
print(housing_results.to_string(index=False))

print("\n Tabla observada:")
print(housing_table.to_string())

print("\n Frecuencias esperadas:")
print(pd.DataFrame(expected_housing,
                   index=housing_table.index,
                   columns=housing_table.columns).round(2).to_string())


 Chi-cuadrado: HOUSING vs IMPAGO

                Métrica                                       Valor
           Chi-cuadrado                                       1.083
                p-value                                    0.298109
     Grados de libertad                                           1
Decisión (alpha = 0.05) No rechazamos H₀ (variables independientes)

 Tabla observada:
credit_default     0   1
housing                 
0               5726  78
1               5101  83

 Frecuencias esperadas:
credit_default        0      1
housing                       
0               5718.96  85.04
1               5108.04  75.96


**Conclusión: No rechazamos la hipótesis nula**

- No existe relación estadísticamente significativa entre tener hipoteca y el impago (p = 0.298).
- Las diferencias entre clientes con y sin hipoteca son pequeñas y compatibles con el azar.
- La tabla observada es muy similar a la esperada, lo que confirma la ausencia de efecto real.

La hipoteca no es un factor de riesgo relevante en el comportamiento de impago. El riesgo está explicado principalmente por el préstamo personal.


